flink-yarn/src/test/java/org/apache/flink/yarn/TestingYarnAMRMClientAsync.java

```diff
@@ -0,0 +1,158 @@
+		super(0, callbackHandler);
+	}
+
+	@Override
+	public List<? extends Collection<AMRMClient.ContainerRequest>> getMatchingRequests(Priority priority, String resourceName, Resource capability) {
+		return getMatchingRequestsFunction.apply(Tuple4.of(priority, resourceName, capability, handler));
+	}
+
+	@Override
+	public void addContainerRequest(AMRMClient.ContainerRequest req) {
+		addContainerRequestConsumer.accept(req, handler);
+	}
+
+	@Override
+	public void removeContainerRequest(AMRMClient.ContainerRequest req) {
+		removeContainerRequestConsumer.accept(req, handler);
+	}
+
+	@Override
+	public void releaseAssignedContainer(ContainerId containerId) {
+		releaseAssignedContainerConsumer.accept(containerId, handler);
+	}
+
+	@Override
+	public void setHeartbeatInterval(int interval) {
+		setHeartbeatIntervalConsumer.accept(interval);
+	}
+
+	@Override
+	public RegisterApplicationMasterResponse registerApplicationMaster(String appHostName, int appHostPort, String appTrackingUrl) {
+		return registerApplicationMasterFunction.apply(appHostName, appHostPort, appTrackingUrl);
+	}
+
+	@Override
+	public void unregisterApplicationMaster(FinalApplicationStatus appStatus, String appMessage, String appTrackingUrl) {
+		unregisterApplicationMasterConsumer.accept(appStatus, appMessage, appTrackingUrl);
+	}
+
+	void setGetMatchingRequestsFunction(
+		Function<Tuple4<Priority, String, Resource, CallbackHandler>, List<? extends Collection<AMRMClient.ContainerRequest>>>
+			getMatchingRequestsFunction) {
+		this.getMatchingRequestsFunction = Preconditions.checkNotNull(getMatchingRequestsFunction);
+	}
+
+	void setAddContainerRequestConsumer(
+		BiConsumer<AMRMClient.ContainerRequest, CallbackHandler> addContainerRequestConsumer) {
+		this.addContainerRequestConsumer = Preconditions.checkNotNull(addContainerRequestConsumer);
+	}
+
+	void setRemoveContainerRequestConsumer(
+		BiConsumer<AMRMClient.ContainerRequest, CallbackHandler> removeContainerRequestConsumer) {
+		this.removeContainerRequestConsumer = Preconditions.checkNotNull(removeContainerRequestConsumer);
+	}
+
+	void setReleaseAssignedContainerConsumer(
+		BiConsumer<ContainerId, CallbackHandler> releaseAssignedContainerConsumer) {
+		this.releaseAssignedContainerConsumer = Preconditions.checkNotNull(releaseAssignedContainerConsumer);
+	}
+
+	void setSetHeartbeatIntervalConsumer(
+		Consumer<Integer> setHeartbeatIntervalConsumer) {
+		this.setHeartbeatIntervalConsumer = setHeartbeatIntervalConsumer;
+	}
+
+	void setRegisterApplicationMasterFunction(
+		TriFunction<String, Integer, String, RegisterApplicationMasterResponse> registerApplicationMasterFunction) {
+		this.registerApplicationMasterFunction = registerApplicationMasterFunction;
+	}
+
+	void setUnregisterApplicationMasterConsumer(
+		TriConsumer<FinalApplicationStatus, String, String> unregisterApplicationMasterConsumer) {
+		this.unregisterApplicationMasterConsumer = unregisterApplicationMasterConsumer;
+	}
```

Comment:
The reason I did not set these functions in the constructor is that, one would have to construct the entire `TestingYarnAMRMClientAsync` for defining one of these functions in a test case. We have several such testing classes (RM client, NM client, slot manager), and it would be complicated if some of the test cases need to provide customized instances for a subset of these classes when creating the `Context`, while some other test cases need to do that for a different subset.

I think making the fields `volatile` sounds good to me.

Reply:
Makes sense.